criblio/appscope
# Enable Integration Tests

This enables running our integration tests for PRs to `master` and `release/*` branches.  Along the way, the tests themselves have been made consistent and accessing for debugging without having to modify them.

Features:
* Run `make` (without a target) in `test/testContainers` to get a list of available targets as well as a list of all of the tests.
  ```text
  $ make
  AppScope Integration Test Runner
    `make help` - show this info
    `make all` - run all tests, stop of first failure
    `make build` - build all test images
    `make push` - push all test images
    `make (test)` - run a single test
    `make (test)-build` - build the test's image
    `make (test)-shell` - run a shell in the test's container
  Tests: alpine bash cli detect_proto elastic gogen go_2 go_3 go_4 go_5 go_6 go_7 go_8 go_9 go_10 go_11 go_12 go_13 go_14 go_15 go_16 java7 java8 java9 java10 java11 java12 java13 java14 kafka logstream nginx oracle-java6 oracle-java7 oracle-java8 oracle-java9 oracle-java10 oracle-java11 oracle-java12 oracle-java13 oracle-java14 splunk syscalls tls transport
  ```
* Run `make all` to manually run all the tests or `make ${TEST}` (i.e. `make alpine`) to run just one.
* Run `make ${TEST}-shell` (i.e. `make alpine-shell`) to start the test container and get to a shell prompt instead of running the test.
* The top-level directory is mounted at `/opt/appscope` when each container is run so you have access to the code and the built (and rebuilt) binaries. There are softlinks to them in `/usr/local/scope/(bin|lib)` too and they're in the `$PATH`.
* A `scope-test` script in every container runs the actual tests. It's in the `$PATH`.
* There should be shell aliases for `l`, `ll`, `ls`, etc. but that's not fully tested on all of them. Same for GDB settings. Some are likely missing tools like gdb, strace, etc. We'll update them as we need to.
* The GitHub workflow will rebuild the test images and push them to Docker Hub for every push to the default branch. The tests are run for non-draft PRs.  

Testing:
* Let them run at GitHub. They should all pass.
* Try `make all`, `make alpine`, and `make alpine-shell`.

To Do:
* The tests are not running as fast as they could be because the existing images are not being use to cache from for some reason. The problem appears to be something specific to Docker in the GitHub Workflow/Action environment. We're going to push this as is and revisit the performance later. When we do so, [this](https://www.docker.com/blog/docker-github-actions/) seems like a good place to start.
* The `kafka` test is still inconsistent. Need to revisit its timing.  

Closes #117.
